### PR TITLE
Feature/apl 703 change genesis shard import

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
added functionality :
* importing data into main db from CSV/zip downloaded from other nodes
* transport subsystem functionality is extended for processing additional request about 'sharding'
* p2p shard archive downloading is added in transport
* node start up logic has changed to wait for information from transport if there is no block data in database
* added code for updating node app status

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+     * @param shardPresentData shard present data contains downloaded ZIP name
+     */
+    public void onShardPresent(@ObservesAsync @ShardPresentEvent(ShardPresentEventType.PRESENT) ShardPresentData shardPresentData) {
+        // shard archive data has been downloaded at that point and stored (unpacked?) in configured folder
+        String zipFileName = shardPresentData.getFileIdValue();
+        Path zipInFolder = csvImporter.getDataExportPath().resolve(zipFileName + ".zip").toAbsolutePath();
+        log.debug("Try unpack file name '{}'", zipInFolder);
+        boolean unpackResult = zipComponent.extract(zipFileName + ".zip", csvImporter.getDataExportPath().toString());
+        log.debug("Zip is unpacked = {}", unpackResult);
+        Genesis.apply(); // import public data - genesis public Keys + balances
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java]: We should import only genesis_public_key

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    public void onNoShardPresent(@ObservesAsync @ShardPresentEvent(ShardPresentEventType.NO_SHARD) ShardPresentData shardPresentData) {
+        // start adding old Genesis Data
+        log.info("Genesis block not in database, starting from scratch");
+        TransactionalDataSource dataSource = databaseManager.getDataSource();
+        Connection con = dataSource.begin();
+        try {
+            Block genesisBlock = Genesis.newGenesisBlock();
+            addBlock(dataSource, genesisBlock);
+            long initialBlockId = genesisBlock.getId();
+            log.debug("Generated Genesis block with Id = {}", initialBlockId);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java]: Looks like this code is copypasted from BlockchainProcessorImpl.addGenesisBlock

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 3]
```diff
@@ -688,7 +688,7 @@ protected int update(int nextUpdate) {
             case 252:
                 // SHARDING meta-info inside main database
                 apply("CREATE TABLE IF NOT EXISTS shard (shard_id BIGINT NOT NULL, shard_hash VARBINARY, " +
-                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY)");
+                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY, generator_ids ARRAY DEFAULT NULL)");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Not null here

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardInitTableSchemaVersion.java [hunk 4]
```diff
@@ -29,7 +29,8 @@ protected int update(int nextUpdate) {
                         "BLOCK_SIGNATURE binary(64) not null, " +
                         "PAYLOAD_HASH binary(32) not null, " +
                         "GENERATOR_ID BIGINT not null, " +
-                        "TIMEOUT INTEGER default 0 not null " +
+                        "TIMEOUT INTEGER default 0 not null, " +
+                        "GENERATOR_IDS ARRAY DEFAULT NULL " +
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardInitTableSchemaVersion.java]: Such code is not needed here

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/mapper/ShardRowMapper.java [hunk 5]
```diff
@@ -15,12 +17,14 @@
...
     public Shard map(ResultSet rs, StatementContext ctx) throws SQLException {
 
+        Long[] generatorIds = DbUtils.getArray(rs, "generator_ids", Long[].class);
         return Shard.builder()
                 .id(rs.getLong("shard_id"))
                 .shardHash(rs.getBytes("shard_hash"))
                 .shardState(rs.getLong("shard_state"))
                 .shardHeight(rs.getInt("shard_height"))
                 .zipHashCrc(rs.getBytes("zip_hash_crc"))
+                .generatorIds(generatorIds == null ? Convert.EMPTY_LONG : Convert.toArray(generatorIds))
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/mapper/ShardRowMapper.java]: Should not be null

Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/mapper/ShardRowMapper.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardInitTableSchemaVersion.java (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardInitTableSchemaVersion.java]: Removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Done - NOT NULL

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java]: **addGenesisBlock()**  was renamed to: 
**private void continuedDownloadOrTryImportGenesisShard()**
Those functionality was moved from BlockchainProcessorImpl to that class.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java]: Done

**Pull request merged**